### PR TITLE
Config refactor

### PR DESCRIPTION
**Pull request by sordonia**
- Configs are now dataclasses
- DataArgs, SelectorArgs, ModifierArgs and TransformArgs automatically read their fields from the registered classes through an ad-hoc metaclass
- This can be considered the first step towards separating arguments such that Trainer and Model itself can use different inits

### CODE REVIEW

File: mttl/config.py [hunk 1]
```diff
@@ -1,89 +1,140 @@
...
-                "The wrapper we provide for flash attention 2 may not behave as expected for"
-                + " some models. Please make sure you test the model with packed sequences"
-            )
+    for f in fields(config_class):
+        if hasattr(args, f.name):
+            value = getattr(args, f.name)
+
+            # if the field is not a multi-default, we override it (it was either default or set by the user)
+            if not isinstance(value, MultiDefaultValue):
+                kwargs[f.name] = value
```
Comment (pclucas14) [inline mttl/config.py:85]: So non overwritten values are always of type `MultiDefaultValue` ?

File: mttl/config.py [hunk 2]
```diff
@@ -1,89 +1,140 @@
...
-        _ = data.pop("_updated_kwargs", None)
-        return cls(kwargs=data, raise_error=False, silent=True)
+    def fromdict(cls, data, strict=False) -> "Args":
+        """
+        Reload a dataclass from a dictionary. We store the class name in the dict to be able to reload it.
+        If the cls is Args, then we try to access the `_class` attribute to get the class name.
+        """
+        cls_name = data.pop("args_class", None)
+        if cls == Args:
+            if not cls_name:
```
Comment (pclucas14) [inline mttl/config.py:108]: when would you do `Args.fromdict` instead of `<subclass>.from_dict` ?

File: mttl/config.py [hunk 3]
```diff
@@ -133,229 +197,321 @@ def save_config(self, output_dir):
...
+    cache_dir: str = os.getenv("CACHE_DIR", "./cache")
+    data_dir: str = os.getenv("TRAIN_DIR", "/tmp/")
+    output_dir: str = os.getenv("OUTPUT_DIR", "./output")
+
+    finetune_task_name: str = None
+    exp_name: str = None
+    expert_name: str = None
+
+    # Training config
+    micro_batch_size: str = None
```
Comment (pclucas14) [inline mttl/config.py:329]: Now would be a good time to clean up the mess that is `micro_batch_size` `train_batch_size` `gradient_accumulation_steps`.

File: projects/modular_llm/playground.py [hunk 4]
```diff
@@ -87,20 +87,19 @@ class ConversationNoTemplate(Conversation):
...
 
-    config = ExpertConfig.parse(
-        raise_error=False,
+    config = MultiExpertModel.parse(
         c="./configs/wiki-mmlu/phi-2_flan_v3.json",
         extra_kwargs=dict(k.split("=") for k in sys.argv[1:]),
     )
     tokenizer = get_tokenizer_with_args(config.model, "gpt", "left", "left", True)
-    model = MultiExpertModel(**vars(config), tokenizer=tokenizer).to("cuda")
+    model = MultiExpertModel(**config.asdict()).to("cuda")
```
Comment (pclucas14) [inline projects/modular_llm/playground.py:102]: does `config` have a pointer to `tokenizer` ? or are we not passing the tokenizer anymore

File: mttl/config.py (hunk 1 above)
Comment (sordonia) [reply mttl/config.py:85]: yeah exactly, given that it's a MultiDefaultValue, we don't pass it to the dataclass i.e. the user has not set it, in this case the dataclass will use its own default

File: mttl/config.py (hunk 2 above)
Comment (sordonia) [reply mttl/config.py:108]: when you reload from the raw dictionary, e.g. maybe some experts have been trained with MoETrainingConfig and you want to inspect some of those argument values

File: mttl/config.py (hunk 3 above)
Comment (sordonia) [reply mttl/config.py:339]: oh, do we have the three of them?

File: projects/modular_llm/playground.py (hunk 4 above)
Comment (sordonia) [reply projects/modular_llm/playground.py:102]: we are not passing tokenizer anymore, afaik it's not needed

**Comment by sordonia** [comment]
Added comments

File: mttl/config.py (hunk 3 above)
Comment (sordonia) [reply mttl/config.py:339]: alloa

**Review by pclucas14 — APPROVED** [review]

**Review by matheper — APPROVED** [review]

File: mttl/config.py (hunk 3 above)
Comment (sordonia) [reply mttl/config.py:347]: `gradient_accumulation_steps` is computed in post_init so it's not settable by the user, it's fine no?

**Pull request merged**